File: 2021/elixir/apps/day7/lib/part2.ex
defmodule Day7Part2 do
  @moduledoc """
  Day7 - Part2
  """

  require Integer

  def solve(input) do
    input
    |> String.split(",")
    |> Enum.map(&String.to_integer/1)
    |> settle()
  end

  # There is in fact a math proof that the answer is either floor(mean) or ceil(mean), though I didn't come up with.
  # see: https://www.reddit.com/r/adventofcode/comments/rawxad/2021_day_7_part_2_i_wrote_a_paper_on_todays
  # or proof_of_part2.pdf
  defp settle(points) do
    all =
      for k <- Enum.min(points)..Enum.max(points),
          do: Enum.reduce(points, 0, fn p, acc -> acc + div(abs(k - p) * (abs(k - p) + 1), 2) end)

    Enum.min(all)
  end
end

File: 2021/go/day7/part2/part2.go
package part2

import (
	"math"
	"strconv"
	"strings"
)

func Solve(input string) string {
	var points []int
	for _, point := range strings.Split(input, ",") {
		n, _ := strconv.Atoi(point)
		points = append(points, n)
	}
	return strconv.Itoa(calc(points))
}

func calc(points []int) int {
	// see proof_of_part2.pdf
	meanFloor, meanCeil := mean(points)
	sum1, sum2 := 0, 0
	for _, point := range points {
		sum1 += abs(point-meanFloor) * (abs(point-meanFloor) + 1) / 2
		sum2 += abs(point-meanCeil) * (abs(point-meanCeil) + 1) / 2
	}
	return min(sum1, sum2)
}

func mean(points []int) (int, int) {
	sum := 0
	for _, point := range points {
		sum += point
	}
	mean := float64(sum) / float64(len(points))
	return int(math.Floor(mean)), int(math.Ceil(mean))
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}
